zebra-chain/src/history_tree.rs

```diff
@@ -0,0 +1,245 @@
+        // The explanation of how it works is from zcashd:
+        // https://github.com/zcash/zcash/blob/0247c0c682d59184a717a6536edb0d18834be9a7/src/coins.cpp#L351
+
+        let mut peak_pos_set = HashSet::new();
+
+        // Assume the following example peak layout with 14 leaves, and 25 stored nodes in
+        // total (the "tree length"):
+        //
+        //             P
+        //            /\
+        //           /  \
+        //          / \  \
+        //        /    \  \  Altitude
+        //     _A_      \  \    3
+        //   _/   \_     B  \   2
+        //  / \   / \   / \  C  1
+        // /\ /\ /\ /\ /\ /\ /\ 0
+        //
+        // We start by determining the altitude of the highest peak (A).
+        let mut alt = (32 - ((self.size + 1) as u32).leading_zeros() - 1) - 1;
+
+        // We determine the position of the highest peak (A) by pretending it is the right
+        // sibling in a tree, and its left-most leaf has position 0. Then the left sibling
+        // of (A) has position -1, and so we can "jump" to the peak's position by computing
+        // -1 + 2^(alt + 1) - 1.
+        let mut peak_pos = (1 << (alt + 1)) - 2;
+
+        // Now that we have the position and altitude of the highest peak (A), we collect
+        // the remaining peaks (B, C). We navigate the peaks as if they were nodes in this
+        // Merkle tree (with additional imaginary nodes 1 and 2, that have positions beyond
+        // the MMR's length):
+        //
+        //             / \
+        //            /   \
+        //           /     \
+        //         /         \
+        //       A ==========> 1
+        //      / \          //  \
+        //    _/   \_       B ==> 2
+        //   /\     /\     /\    //
+        //  /  \   /  \   /  \   C
+        // /\  /\ /\  /\ /\  /\ /\
+        //
+        loop {
+            // If peak_pos is out of bounds of the tree, we compute the position of its left
+            // child, and drop down one level in the tree.
+            if peak_pos >= self.size {
+                // left child, -2^alt
+                peak_pos -= 1 << alt;
+                alt -= 1;
+            }
+
+            // If the peak exists, we take it and then continue with its right sibling.
+            if peak_pos < self.size {
+                // There is a peak at index `peak_pos`
+                peak_pos_set.insert(peak_pos);
+
+                // right sibling
+                peak_pos = peak_pos + (1 << (alt + 1)) - 1;
+            }
+
+            if alt == 0 {
+                break;
+            }
+        }
+
+        // Remove all non-peak entries
+        self.peaks.retain(|k, _| peak_pos_set.contains(k));
+        // Rebuild tree
+        self.inner = Tree::new_from_cache(
+            self.network,
+            self.network_upgrade,
+            self.size,
+            &self.peaks,
+            &Default::default(),
+        )?;
```

Comment:
Oh so this just gets blown away and replaced? Do we have to double check `height` and `size` of `HistoryTree` when we do this?

Reply:
Exactly! We _can_ keep the same tree around without pruning, either because of efficiency or to allow node removal (see #2378), but teor suggested keeping things simple for now.

Sorry, I don't understand, what do you mean by double checking those?

Reply:
Ah we use our `self.size` to reconstruct `InnerHistoryTree::new_from_cache`, nevermind about `self.current_height` 👍